### ISSUE TITLE
"optimizations"

### ISSUE THREAD

**Issue by AeroNotix** [ibnfirnas/dups#1]
Hi Siraaj! 

Crazy to see you working on this. I've been playing with creating the most optimal "algorithm" to find duplicated files myself.

My issue: I have an overly zealous backer-upper wife who duplicates her 3D models over and over again. I have around 2TB of total disk space of which 35% is used. Many files are multiple gigabytes in size, and many are smaller (jpg, png etc). All of the open source tools were far too slow. Some were decent but I looked into making it faster, here's my implementation: https://github.com/AeroNotix/fstdupe

I initially wanted to implement it like you have, with shell and piped commands to each other but I couldn't get the performance I wanted (though a pure shell implementation would be very "pure" :)) 

I find this problem deceptively deep. There are many factors that go into making the most performant solution. The kinds of files that are duplicate, the number of files that are duplicates, the depth, the disk type (SSD, HDD), filesystem, processor speed (threads, clock speed etc) all come into play. I've come to the conclusion that there is no single method to be the most performant. Any implementation should be able to switch between different techniques based on attributes of the files being processed. 

Any way, some small insights into how I made this implementation the fastest I could find on the internet:

* Group files first by file size. Most files fall into two categories, zero byte sized files and N byte sized files. As you well know, there are many numbers and most of them are different :) most files occupy different sizes. Obvious but true. Zero byte sized files are ignored in my implementation as all I am trying to do is recover unnecessarily used disk space. If there is a byte size with only one file, there exists no duplicate. If there are files with exact byte size matches they possibly are duplicate.
* Chunk the filepaths into N buckets based on some factor (I chose CPU). I'm aware there's no real direct connection between available processors and disk performance but it netted me the best performance. Run through buckets using a different core each.
* Use the quickest hash possible. MD5 is good, but slow compared to the very simple CRC.
* Read a small portion of the file, this saves on hashing the entire file. Optionally hash this small chunk or just compare directly. I found not hashing this small chunk gave the best time savings versus CPU/memory. If they are the same, they potentially are duplicated, also potentially not.
* Once CRC'd and potential duplicates are located, hash (optionally with a hash less prone to collisions) them in their entirety again and present the final set of files to the user.

Essentially all these steps filter out files which definitely do not have duplicates the quickest way. You don't end up hashing every file unnecessarily.

For me and what I assume is the potential use-case for a lot of these duplicate file finders this algorithm is the most performant. I've compared my implementation to the most popular open source implementations and mine performs the best for my data set.

That said, I know with this algorithm you could construct a set of files to completely undermine the performance. I believe this is not a failure in the algorithm but in the fact that there is no single best-performing algorithm in all use-cases. This is why I mentioned that the implementation should optionally be instructed to behave differently based on what the user understands about the data.

I.e. Lots of similarly sized files but a low number of duplicates? Skip checking file sizes and go straight to hashing. Lots of files with a similar prefix? Skip checking prefixes and go straight to hashing.

Anyway, I just found it very interesting you also were working on this and wanted to share my thoughts. When I initially implemented this I thought it trivial but the more I thought about it I discovered it was a rich vein of masturbatory science projectery. 

Thanks!

**Comment by xandkar** [comment]
Hello there, Aaron!

These are excellent insights - thank you!!!

Like you, I expected to be done with this after 10-30 minutes of `bash`ing, but there's just so many (reasonable, but yet unexpected) issues that keep popping-up! My personal biggest annoyance was file names with spaces, quotes and newlines, which I could not easily parse in AWK (and which do not need to be parsed at all if using an API which isn't just a giant string...)

The lowest hanging nuts here are the 0-sized files, so I'll adapt skipping those ASAP. Yeah, technically those have the same content, but like you said - we wouldn't be using anything beyond `find -size 0` if that is what we wanted to find...

I really, really love the idea of using the progressively stronger measures of equality!
1. size;
2. small content sample(s);
3. weak hashing;
4. stronger hashing (possible multiple levels here too);
5. perhaps (optionally) culminating with a `diff` for some files?

This I will do 2nd and parallelization probably last.

Again, thank YOU for the invite to this hidden gem of a circle jerk!

**Comment by AeroNotix** [comment]
Yo! 

@puzza007  was mentioning trying it out with Rust for better CPU performance along with a neat work stealing thread pool library. It would help keep cores active versus my 'chunked work pool' implementation.

I had some ideas around maintaining file descriptors open and passing those fd's between the various filter steps to minimize I/O though I am yet to find a clean solution doing that, especially so when there are a large number of files to check.

Also, there could be a good speed up for spinning rust disks by somehow ordering the reads based on where they are placed physically to minimize random accesses. I am not sure exactly how to manage that, would be interested if you think that would provide any improvement. That said, ordering the reads by physical location may interfere with any parallelised steps later, though I am sure it will reduce overall random I/O.

Finally, if all this could be achieved by simple shell commands piped to each other I would be very impressed. It seems that it really should be possible but needs more thought.

**Comment by xandkar** [comment]
So I just implemented skipping uniquely-sized files and that alone yielded a ~3x speedup!

**Comment by AeroNotix** [comment]
Nice! Really interested in how far this can go.

**Comment by gornik** [comment]
Hi guys, hope I'm not late to the party. That's a really interesting problem, putting some random thoughts here.

The way I see it the goal is to prove that in a set of files a given file is unique, so that we can discard it quickly, some ways to achieve that:
 * compare sizes
 * probe random bytes from two files, probably not from the beginning (because file headers might be the same etc)
 * start reading the files by blocks of a given size, calculate a fast rolling hash for the blocks, if at any point the hashes differ, the files are different
 * calculate a full hash with a very low probability of collisions.

Check out how rsync does things BTW, it's a similar problem.

What else you could parallelize/optimize (not sure if this would actually speed things up, but may be worth a try)?
* the initial directory walk:
  * given a directory tree to check, you can start walking it and gathering files, it would be hard to do it in parallel, but listing directories is usually very fast, so the speedup won't be that great
  * you don't need to wait for this step to finish, once the first files show you you can start streaming them to the next steps
* gathering files sizes / grouping:
  * now this can totally be done in parallel, but since stat is usually fast it may not be worth it
  * once you have the size, you can start partitioning further work by the size, e.g. have a pool of workers that do the grouping by size
* probing / rolling hash / full hash:
  * you can start calculating these as soon as you have two suspects (e.g. two files with the same size)
  * if you start without waiting for a full set of files, you might need to keep the results if the next file shows up, example: you've got files A and B, you start calculating a hash for each block, the files are different at block 56, but you can't discard the results, since another file C might show up and it might be a duplicate of A or B, if you want to discard C quickly, you'll need to keep the previous hashes for A and B for all 56 blocks and if C is the same as A, you might need to resume comparing blocks for A. 
  * discarding files that are smaller than x bytes would probably help a lot: there is usually lot of them, but we mostly care about big files

Here is my initial take on this in Elixir using Flow, although lots of work still needs to be done there:
https://gist.github.com/gornik/2704b543740c1d573e8b2e7d4a84cf02

**Comment by AeroNotix** [comment]
So far I think we've settled on xxhash being the fastest overall. That said, I want to investigate if different hashes perform better with different input sizes.

I haven't looked deeply into the internals of rsync, though I use it quite a bit and I find that its performance is quite bad realistically to what should theoretically be possible, especially if you're rsyncing between two/more local disks. Over the network is a whole different thing.

[In my implementation](https://github.com/AeroNotix/fstdupe) all steps are in parallel. That means the directory walk, the file size grouping and hashing steps are all done in parallel and speed ups were found every time something became parallelised. A note about this implementation: compared to the Rust implementation (it's on @puzza007's private bitbucket I believe, so might not be able to link). The performance is equal only if I disable the GC entirely in the Go implementation.

Streaming the files between steps probably is a good idea as there really isn't a need to fully complete each step before the next begins. I was thinking of doing something along the lines of, if a step discovers potential duplicates for a given set of files, kick off the next step for that set of files or have a group of workers consume a queue of jobs and their jobs can be triggered when two jobs come in for that potential duplicate.

If there exists *two* duplicates for a given size/initial hash then keeping around results in case another file comes in isn't any additional work, so again, worth looking into.

As for checking random blocks to reduce the possibility of certain types of files having duplicate headers but remaining bodies not, it's a good idea to look into but I wonder if the file open/seek/compare steps might incur additional overhead compared to a naive check of the initial blocks. Would be good to know what are the average potentially duplicated header sizes for a few popular formats and choose an initial comparison size based on that. I know that doesn't help for custom file formats but would give a good ball park initial compare size that covers a lot of bases.

Additionally, the biggest "resource" that ends up working at odds with parallelising tasks I've found is actually available file handles. You can write your algorithm to consume as many as you can and perform rolling hashes with those and even pass the fd's between steps but it fails once you have a larger number of files to check. I experimented with various caching techniques and found none yielded any improvements.

**Comment by gornik** [comment]
@AeroNotix Amazing work! I've checked your implementation and it's blazing fast.

**Comment by puzza007** [comment]
> [In my implementation](https://github.com/AeroNotix/fstdupe) all steps are in parallel. That means the directory walk, the file size grouping and hashing steps are all done in parallel and speed ups were found every time something became parallelised. A note about this implementation: compared to the Rust implementation (it's on @puzza007's private bitbucket I believe, so might not be able to link). The performance is equal only if I disable the GC entirely in the Go implementation.

I've made my repo public (https://bitbucket.org/puzza007/ksero/src/master/). It's basically a single threaded approach with [Rayon's](https://github.com/rayon-rs/rayon) `par_iter()` sprinkled on top. Fun problem!

**Comment by AeroNotix** [comment]
Cheers!
